[jira] Verify that projects referenced by issues exist

sochotnicky opened the pull request:
This is a bigger change that should probably be split into smaller steps.
Changes:
 * Adds 3 new env variables for base url, username and password
 * Actually use JIRA API (with basic auth) to get list of actual projects
 * Only output JIRA issues if their projects exist in JIRA
 * Use alternative passive method registration to allow multiple issues being
   mentioned in single message
 * Tweak regex pattern so only upper-caps are taken into account but allow
   non-whitespace preceding characters

Maybe alternative "simple" mode could be preserved - for example if
JIRA_BASE_URL is not set, keep working in the old way?

This also needs tweaks to tests (tbd depending if the general direction is OK).

I have further changes planned wrt actually being able to output summary and other info from the issues to the channels instead of just simple link

Related to #33

Commit:
Handle multiple issues in single message

This provides proper handling of multiple issues in single message. We
use different register command so we can reply with multiple replies

fabioxgn commented on jira/jira.go:
```diff
@@ -1,41 +1,101 @@
+
+	client, err = gojira.NewClient(tp.Client(), baseURL)
+	if err != nil {
+		panic(err)
```
@sochotnicky this probrably would prevent the bot from booting up. I think it's a better ideia to just log the error and do not register the command if the initialization fails.

sochotnicky replied:
I can tweak though I actually did this on purpose, thinking that if this fails the configuration is seriously borked and it's better to crash during startup rather than fail during operation (i.e. fail fast). 

I will change it

fabioxgn commented on jira/jira.go:
```diff
@@ -1,41 +1,101 @@
-	if re.MatchString(text) {
-		return re.FindString(text)
+func getProjects() map[string]gojira.Project {
+	req, _ := client.NewRequest("GET", "rest/api/2/project", nil)
```
It's better to always handle the errors, or thinks might get very hard to debug later.

sochotnicky replied:
yeah, you are right (part of WIP I guess) :-)

fabioxgn commented on jira/jira.go:
```diff
@@ -1,41 +1,101 @@
+	_, err := client.Do(req, projectObjects)
+	if err != nil {
+		panic(err)
+	}
```
if you can't process the request, return an error and let the caller handle it, panic is not a good idea.

sochotnicky replied:
KInda similar to comment above, since this is called during startup I though panic might be better than letting people debug why the bot is not responding to any JIRA mentions. Logging the error should help with that though so I'll tweak

sochotnicky commented:
Any thoughts on whether I should try to keep the old unauthenticated/simple way working for the future?

Commit:
Query JIRA with basic auth to verify projects exist

With this change the bot will not respond to bogus mentions of issues
if their projects do not actually exist in configured JIRA instance.
This check is done using basic auth and JIRA api query.

This needs 3 additional env variables to be set up:
 * JIRA_BASE_URL
 * JIRA_USER
 * JIRA_PASS

sochotnicky commented:
OK, so made the error handling tweaks, added a few tests (though I can think of more to add later). 

For now when the JIRA querying during init fails - code does not register the passive command. I could tweak so that it would keep working in the old way as a fallback?

Commit:
Add JIRA plugin readme file

fabioxgn commented on jira/jira.go:
```diff
@@ -1,41 +1,110 @@
+func getProjects() (map[string]gojira.Project, error) {
+	req, err := client.NewRequest("GET", "rest/api/2/project", nil)
+	if err != nil {
+		return projects, errors.New("Error creating request object: " + err.Error())
```
You can replace this with:

```Go
return fmt.Errorf("Error creating request object: %v", err)
```

fabioxgn commented on jira/jira.go:
```diff
@@ -1,41 +1,110 @@
+	projects = make(map[string]gojira.Project)
+	_, err = client.Do(req, projectObjects)
+	if err != nil {
+		return projects, errors.New("Failed getting JIRA projects: " + err.Error())
```
Can replace this with `fmt.Errorf` too.

fabioxgn commented on jira/jira.go:
```diff
@@ -1,41 +1,110 @@
+
+	_, err = getProjects()
+	if err != nil {
+		log.Printf("Error querying JIRA for projects: " + err.Error())
```
You can use the format insteat of `+`

```Go
log.Printf("Error querying JIRA for projects: %v\n", err))
```

Also notice that Printf doesn't add new lines, so always end it with a `\n`

fabioxgn commented:
> Any thoughts on whether I should try to keep the old unauthenticated/simple way working for the future?

About this, I don't think we should keep the old one, it is so simple that I doubt that anyone use it, and if it does it's just a couple of lines of code to restore.

Commit:
Remove original env variable and update main README

We point in jira specific readme for configuration information. Also got
rid of original env variable - the URL is now automatically created from
base URL env variable
